### PR TITLE
ASoC/soundwire: Add rt712 VB support

### PR DESCRIPTION
**Pull request by bardliao**
Lookup SDCA functions and version to identify rt712/rt713 B version. This PR is from https://github.com/thesofproject/linux/pull/5128

### CODE REVIEW

**Review by vijendarmukunda — APPROVED** [review]
LGTM

**Comment by vijendarmukunda** [comment]
@bardliao : I do have one question. When can we expect SDCA 1.0 migration support related patches going to land in kernel mainline. Any ETA planned for the same? 
We also want to validate these changes with AMD SoundWire stack.

**Comment by bardliao** [comment]
> @bardliao : I do have one question. When can we expect SDCA 1.0 migration support related patches going to land in kernel mainline. Any ETA planned for the same? We also want to validate these changes with AMD SoundWire stack.

Sorry @vijendarmukunda I didn't get your question. What are "SDCA 1.0 migration support related patches"? Is there anything we missed for supporting SDCA codecs?

**Comment by vijendarmukunda** [comment]
> > @bardliao : I do have one question. When can we expect SDCA 1.0 migration support related patches going to land in kernel mainline. Any ETA planned for the same? We also want to validate these changes with AMD SoundWire stack.
> 
> Sorry @vijendarmukunda I didn't get your question. What are "SDCA 1.0 migration support related patches"? Is there anything we missed for supporting SDCA codecs?

@bardliao : As per my understanding, Current codec drivers supports SDCAv0.6 version.
If I understand correctly SDCAv1.0 expects all descriptors and peripheral topology entries will be part of ACPI tables.
Even peripheral register blind write table will also be part of ACPI table unlike SDCAv0.6 version based codec driver where tables are listed in driver source code. 
As you are working on adding sdca helpers to read from ACPI tables, i thought going forward peripheral devices will conformant with SDCA 1.0.
Please correct me, if my understanding is wrong.

**Comment by charleskeepax** [comment]
I think perhaps that sort of framing is coming more from the Windows side of the world. From the Linux kernel side we should just be supporting both versions as there are parts in the wild using both. The new code being added here doesn't really change any of the existing support however and there is basically no SDCA support currently in the kernel, this is really the first part.

**Comment by bardliao** [comment]
> > > @bardliao : I do have one question. When can we expect SDCA 1.0 migration support related patches going to land in kernel mainline. Any ETA planned for the same? We also want to validate these changes with AMD SoundWire stack.
> > 
> > 
> > Sorry @vijendarmukunda I didn't get your question. What are "SDCA 1.0 migration support related patches"? Is there anything we missed for supporting SDCA codecs?
> 
> @bardliao : As per my understanding, Current codec drivers supports SDCAv0.6 version. If I understand correctly SDCAv1.0 expects all descriptors and peripheral topology entries will be part of ACPI tables. Even peripheral register blind write table will also be part of ACPI table unlike SDCAv0.6 version based codec driver where tables are listed in driver source code. As you are working on adding sdca helpers to read from ACPI tables, i thought going forward peripheral devices will conformant with SDCA 1.0. Please correct me, if my understanding is wrong.

I got your question now @vijendarmukunda. Yes, we are working on SDCA library. However, there is no ETA so far. The SDCA 1.0 codecs will use their own codec drivers like what we do today. We could work together on the SDCA library :)

**Review by shumingfan — APPROVED** [review]

**Comment by bardliao** [comment]
SOFCI TEST

**Comment by vijendarmukunda** [comment]
> > > > @bardliao : I do have one question. When can we expect SDCA 1.0 migration support related patches going to land in kernel mainline. Any ETA planned for the same? We also want to validate these changes with AMD SoundWire stack.
> > > 
> > > 
> > > Sorry @vijendarmukunda I didn't get your question. What are "SDCA 1.0 migration support related patches"? Is there anything we missed for supporting SDCA codecs?
> > 
> > 
> > @bardliao : As per my understanding, Current codec drivers supports SDCAv0.6 version. If I understand correctly SDCAv1.0 expects all descriptors and peripheral topology entries will be part of ACPI tables. Even peripheral register blind write table will also be part of ACPI table unlike SDCAv0.6 version based codec driver where tables are listed in driver source code. As you are working on adding sdca helpers to read from ACPI tables, i thought going forward peripheral devices will conformant with SDCA 1.0. Please correct me, if my understanding is wrong.
> 
> I got your question now @vijendarmukunda. Yes, we are working on SDCA library. However, there is no ETA so far. The SDCA 1.0 codecs will use their own codec drivers like what we do today. We could work together on the SDCA library :)

@bardliao: Thanks

File: sound/soc/sdca/sdca_device.c [hunk 1]
```diff
@@ -22,3 +23,40 @@ void sdca_lookup_interface_revision(struct sdw_slave *slave)
...
+		    id->part_id != 0x716 &&
+		    id->part_id != 0x717)
+			return false;
+
+		for (i = 0; i < slave->sdca_data.num_functions; i++) {
+			if (slave->sdca_data.sdca_func[i].type ==
+			    SDCA_FUNCTION_TYPE_SMART_MIC)
+				return true;
+		}
+
```
Comment (charleskeepax) [inline sound/soc/sdca/sdca_device.c]: Probably be nice to leave the switch but factor this code out into a function like sdca_device_quirk_rt712_vb or something, that is called from here. Whilst this looks fine now whilst there is one quirk this function will get crazy fast when there are a lot of quirks.

Comment (charleskeepax) [reply sound/soc/sdca/sdca_device.c]: Otherwise I think the rest of this looks good to me.

Comment (bardliao) [reply sound/soc/sdca/sdca_device.c]: > Probably be nice to leave the switch but factor this code out into a function like sdca_device_quirk_rt712_vb or something, that is called from here. Whilst this looks fine now whilst there is one quirk this function will get crazy fast when there are a lot of quirks.

Done

**Review by shumingfan — APPROVED** [review]

File: sound/soc/intel/common/soc-acpi-intel-mtl-match.c [hunk 2]
```diff
@@ -9,7 +9,10 @@
 #include <sound/soc-acpi.h>
 #include <sound/soc-acpi-intel-match.h>
 #include <sound/soc-acpi-intel-ssp-common.h>
+#include <sound/sdca.h>
+#include <linux/soundwire/sdw_intel.h>
```
Comment (ujfalusi) [inline sound/soc/intel/common/soc-acpi-intel-mtl-match.c]: nitpick: I would have added these in alphabetic order.

File: sound/soc/intel/common/soc-acpi-intel-sdca-quirks.c [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+/*
+ * soc-acpi-intel-sdca-quirks.c - tables and support for SDCA quirks
+ *
+ * Copyright (c) 2024, Intel Corporation.
+ *
+ */
+
+#include <sound/soc-acpi.h>
+#include <sound/sdca.h>
+#include <linux/soundwire/sdw_intel.h>
```
Comment (ujfalusi) [inline sound/soc/intel/common/soc-acpi-intel-sdca-quirks.c:9]: nitpick: Alphabetic order?

File: sound/soc/sdca/sdca_functions.c [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+
+/*
+ * The MIPI SDCA specification is available for public downloads at
+ * https://www.mipi.org/mipi-sdca-v1-0-download
+ */
+
+#include <linux/acpi.h>
+#include <linux/soundwire/sdw.h>
+#include <sound/sdca_function.h>
+#include <sound/sdca.h>
```
Comment (ujfalusi) [inline sound/soc/sdca/sdca_functions.c:11]: nitpick: in sdca_device.c the order is different:
```c
#include <linux/acpi.h>
#include <linux/soundwire/sdw.h>
#include <sound/sdca.h>
#include <sound/sdca_function.h>
```

**Review by ujfalusi — DISMISSED** [review]
No functional objections, but some nitpicks on header include orders.

**Comment by bardliao** [comment]
@ujfalusi header include orders fixed.

**Review by ujfalusi — APPROVED** [review]

**Pull request merged**